### PR TITLE
Tools: start of script to help decode stm32 ICSR register

### PR DESCRIPTION
**Pull request by peterbarker**
```
pbarker@bluebottle:~/rc/ardupilot(master)$ ./Tools/scripts/decode-ICSR.py 4196355
VECTACTIVE: 3  (Hard fault)
RESERVED1: 0
ROTOBASE: 1
VECTPENDING: 0
RESERVED2: 0
ISRPENDING: 1
RESERVED3: 0
PENDSTCLR: 0
PENDSVCLR: 0
PENDSVSET: 0
RESERVED4: 0
NMIPENDSET: 0
pbarker@bluebottle:~/rc/ardupilot(master)$
```

### CODE REVIEW

File: Tools/scripts/decode-ICSR.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        13: "Reserved",
+        14: "PendSV",
+        15: "SysTick",
+    }
+    if value in exceptions:
+        exception = "%s" % str(exceptions[value])
+    else:
+        exception = "IRQ%u" % (value - 16)
+
+    print(" (%s)" % exception),
```
Comment (khancyr) [inline Tools/scripts/decode-ICSR.py]: trailling comma here

File: Tools/scripts/decode-ICSR.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    except ValueError:
+        return int(s, 16)
+
+
+parser = optparse.OptionParser(__file__)
+
+opts, args = parser.parse_args()
+
+if len(args) == 0:
+    print(parser.usage)
```
Comment (khancyr) [inline Tools/scripts/decode-ICSR.py]: usage is missing !

**Comment by khancyr** [comment]
on python3, the output got many line return between line

File: Tools/scripts/decode-ICSR.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        start_bit = int(bits)
+        stop_bit = int(bits)
+    mask = 0
+    for i in range(start_bit, stop_bit+1):
+        mask |= 1<<i
+    value = (ICSR & mask) >> start_bit
+    print("%s: %u" % (name, value)),
+    if decoder is not None:
+        decoder(value)
+    print("")
```
Comment (khancyr) [inline Tools/scripts/decode-ICSR.py]: this one seems unecessary

File: Tools/scripts/decode-ICSR.py (hunk 1 above)
Comment (peterbarker) [reply Tools/scripts/decode-ICSR.py]: In Python2 the trailing comma mean, "don't add a newline".

That appears to have gone away in Python3, as you noted elsewhere.

File: Tools/scripts/decode-ICSR.py (hunk 2 above)
Comment (peterbarker) [reply Tools/scripts/decode-ICSR.py]: It's generated.  I don't really want to fix the usage *right* now - really want to work out how to specify the arch before then.

File: Tools/scripts/decode-ICSR.py (hunk 3 above)
Comment (peterbarker) [reply Tools/scripts/decode-ICSR.py]: It is - to add the newline.  I've changed the other prints to `sys.stdout.write`

**Pull request merged**